crates/pathfinder/src/merkle_tree/storage.rs

```diff
@@ -0,0 +1,676 @@
+                }
+                PersistedNode::Leaf => {}
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns the node given by `key`, or [None] if it doesn't exist.
+    pub fn get(&self, key: StarkHash) -> anyhow::Result<Option<PersistedNode>> {
+        let hash = key.to_be_bytes();
+
+        let node = self
+            .transaction
+            .query_row(
+                &format!("SELECT data FROM {} WHERE hash = :hash", &self.table),
+                named_params! {
+                    ":hash": &hash[..],
+                },
+                |row| {
+                    let data: Vec<u8> = row.get("data")?;
+                    Ok(PersistedNode::deserialize(&data))
+                },
+            )
+            .optional()?;
+
+        node.transpose()
+    }
+
+    /// Deletes the given node from storage, and decrements the reference count of the node's
+    /// children (if any).
+    ///
+    /// If a child's reference count reaches 0 as a result of this operation,
+    /// it will get deleted in turn. A single deletion may therefore result in a daisy-chain of
+    /// deletions.
+    ///
+    /// ### Warning
+    ///
+    /// Does not perform rollback on failure. This implies that you should rollback the [RcNodeStorage's](RcNodeStorage) transaction
+    /// if this call returns an error to prevent database corruption.
+    fn delete_node(&self, key: StarkHash) -> anyhow::Result<()> {
+        let hash = key.to_be_bytes();
+
+        let node = match self.get(key)? {
+            Some(node) => node,
+            None => return Ok(()),
+        };
+
+        self.transaction.execute(
+            &format!("DELETE FROM {} WHERE hash = :hash", &self.table),
+            named_params! {
+               ":hash": &hash[..],
+            },
+        )?;
+
+        match node {
+            PersistedNode::Binary(binary) => {
+                self.decrement_ref_count(binary.left)?;
+                self.decrement_ref_count(binary.right)?;
+            }
+            PersistedNode::Edge(edge) => self.decrement_ref_count(edge.child)?,
+            PersistedNode::Leaf => {}
+        }
+
+        Ok(())
+    }
+
+    /// Decrements the reference count of the node and automatically deletes it
+    /// if the count becomes zero.
+    pub fn decrement_ref_count(&self, key: StarkHash) -> anyhow::Result<()> {
+        let hash = key.to_be_bytes();
+
+        let ref_count = self
+            .transaction
+            .query_row(
+                &format!("SELECT ref_count FROM {} WHERE hash = :hash", &self.table),
+                named_params! {
+                    ":hash": &hash[..],
+                },
+                |row| {
+                    let ref_count: u16 = row.get("ref_count")?;
+
+                    Ok(ref_count)
+                },
+            )
+            .optional()?;
+
+        match ref_count {
+            Some(0 | 1) => self.delete_node(key)?,
```

Comment:
I'm wondering if I understood this correctly:
- refcount of `0` means: construction in progress, my parent hasn't been added (yet) or I'm root and the tree hasn't been `commit()`-ed,
- refcount of `1` means: constructed, my parent has been added, the tree has been commited if I'm root,
- the case for `Some(0)` here is a failsafe for mid-constructed trees - for example in tests - right?

Reply:
Its also a safeguard against a higher up caller deleting the same tree multiple times.

Once we get around to logging I want to split that to issue a warning on `Some(0)` (but I don't want to crash).

Although maybe one can argue that crashing early on is better as a sign of a bug.

Reply:
But yes you have the correct idea!

Reply:
Awesome! :pray: :tada: